_posts/2017-11-07-haskell-learning-group-our-progress-and-updates.md

```diff
@@ -0,0 +1,71 @@
+
+One thing we've noticed and that was to expect, is that the difficulty of a content is slowly increasing. We are spending more time reading chapters. Sometimes the same material has to be read multiple times in order to fully understand what is it about.
+
+It helps a lot that the most of the concepts are familiar to us through other languages or libraries we are using. Let's say JavaScript doesn't support currying natively the way Haskell does it, but there are libraries which introduce these concepts. We are very often using [RamdaJS][ramdajs] in our projects, so such things as function composition, currying, partial application and so on were not something totally new to us. They were still had to be "translated" and demonstrated in a Haskell way but we certainly didn't need much time to get the idea.
+
+Similar situation was with typing system and type definitions. As many people from the group doing i.e. TypeScript or Scala, there were not that many surprises for us. Nevertheless it's pretty impressive what can be achieved with type constraints in Haskell.
+
+[ramdajs]: http://ramdajs.com
+
+## Types
+
+We went through many topics in the book and since the last update we've covered chapter four, five and discussed a half of the chapter six which covers Typeclasses. The main topics were:
+* Basic DataTypes
+* Types
+* Typeclasses
+
+We learned important basics that helps a lot when you read and try to understand what certain code does. For instance understanding how to read type signature, so something like this starting to make sense:
+
+`(Num a, Num b) => a -> b -> b`
+
+In this case there are two constraints, one for each parameter. They must be of a typeclass Num. You might already have some idea what function might do, based on its type signature.
+We had a look on different examples of how type signatures limit a scope of what body of the function might be. Even more, some  type signatures of function tells us exactly what it does because its the only possible implementation of the function. The simplest example of that will be an identity function:
+
+```haskell
+id :: a -> a
+id x = x
+```
+
+The only thing this function can possibly do is return parameter that was passed to it. More advanced example when there is only one possible implementation is function composition:
+
+```haskell
+f :: (b -> c) -> (a -> b) -> (a -> c)
+f x y = \z -> x(y(z))
+```
+
+It's not obvious from the first look at this type definition that there is no other possible implementation of a function based on its type. Compiler won't let us implement it in any other way. 
```

Comment:
I don't know if I wouldn't stop at just giving `(Num a, Num b) => a -> b -> b` example w/o explaining this one and any further sample and complex type signature above.

Reply:
I understand your point, I was questioning this myself too. Last time I was criticized for explaining too much though

Reply:
imho an article on this blog can be multi-purposes: recap (for/by Sergii) learning, teaching, summarizing and last but not least showing what cool stuff we do at HC.
And it has Sergii's touch, so I like the not pure focus on one thing. Which of course also should not prevent anyone from writing a pure "educational" article. #my2cents
Really cool effort Sergii!

Reply:
Effort is indeed cool :) It's just matter of to what we did agree with this blog post series - no one (I think) would stop @HeeL from adding new blogpost but from `how we learn` series :)